src/gui/vectortile/qgsvectortilebasicrendererwidget.cpp

```diff
@@ -0,0 +1,467 @@
+{
+  Q_UNUSED( column )
+
+  if ( action == Qt::IgnoreAction )
+    return true;
+
+  if ( !data->hasFormat( QStringLiteral( "application/vnd.text.list" ) ) )
+    return false;
+
+  if ( parent.column() > 0 )
+    return false;
+
+  QByteArray encodedData = data->data( QStringLiteral( "application/vnd.text.list" ) );
+  QDataStream stream( &encodedData, QIODevice::ReadOnly );
+  int rows = 0;
+
+  if ( row == -1 )
+  {
+    // the item was dropped at a parent - we may decide where to put the items - let's append them
+    row = rowCount( parent );
+  }
+
+  while ( !stream.atEnd() )
+  {
+    QString text;
+    stream >> text;
+
+    QDomDocument doc;
+    if ( !doc.setContent( text ) )
+      continue;
+    QDomElement rootElem = doc.documentElement();
+    if ( rootElem.tagName() != QLatin1String( "vector_tile_basic_renderer_style_mime" ) )
+      continue;
+
+    QgsVectorTileBasicRendererStyle style;
+    style.readXml( rootElem, QgsReadWriteContext() );
+
+    insertStyle( row + rows, style );
+    ++rows;
+  }
+  return true;
+}
+
+
+//
+
+
+QgsVectorTileBasicRendererWidget::QgsVectorTileBasicRendererWidget( QgsVectorTileLayer *layer, QgsMapCanvas *canvas, QgsMessageBar *messageBar, QWidget *parent )
+  : QgsMapLayerConfigWidget( layer, canvas, parent )
+  , mVTLayer( layer )
+  , mMessageBar( messageBar )
+{
+  setupUi( this );
+  layout()->setContentsMargins( 0, 0, 0, 0 );
+
+  if ( layer->renderer() && layer->renderer()->type() == QStringLiteral( "basic" ) )
+  {
+    mRenderer.reset( static_cast<QgsVectorTileBasicRenderer *>( layer->renderer()->clone() ) );
+  }
+  else
+  {
+    mRenderer.reset( new QgsVectorTileBasicRenderer() );
+  }
+
+  mModel = new QgsVectorTileBasicRendererListModel( mRenderer.get(), viewStyles );
+  viewStyles->setModel( mModel );
+
+  connect( btnAddRule, &QPushButton::clicked, this, &QgsVectorTileBasicRendererWidget::addStyle );
+  connect( btnEditRule, &QPushButton::clicked, this, &QgsVectorTileBasicRendererWidget::editStyle );
+  connect( btnRemoveRule, &QAbstractButton::clicked, this, &QgsVectorTileBasicRendererWidget::removeStyle );
+
+  connect( viewStyles, &QAbstractItemView::doubleClicked, this, &QgsVectorTileBasicRendererWidget::editStyleAtIndex );
+
+  connect( mModel, &QAbstractItemModel::dataChanged, this, &QgsPanelWidget::widgetChanged );
+  connect( mModel, &QAbstractItemModel::rowsInserted, this, &QgsPanelWidget::widgetChanged );
+  connect( mModel, &QAbstractItemModel::rowsRemoved, this, &QgsPanelWidget::widgetChanged );
+}
+
+QgsVectorTileBasicRendererWidget::~QgsVectorTileBasicRendererWidget() = default;
+
+void QgsVectorTileBasicRendererWidget::apply()
+{
+  mVTLayer->setRenderer( mRenderer->clone() );
+}
+
+void QgsVectorTileBasicRendererWidget::addStyle()
+{
+  QStringList lst;
+  lst << tr( "Marker" ) << tr( "Line" ) << tr( "Fill" );
+  QString type = QInputDialog::getItem( this, tr( "Add style" ), tr( "Please choose symbol type" ), lst, 0, false );
```

Comment:
Would be nicer UI to make btnAddRule a QToolButton and put this choice in a popup menu

Reply:
Good point - updated!